### PR TITLE
Onboard seq2seq inputs to new object

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Part of #964 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/core/data/new_data_module.py [hunk 1]
```diff
@@ -290,27 +295,20 @@ def _predict_dataloader(self) -> DataLoader:
...
-        for state in input._state.values():
-            data_pipeline_state.set_state(state)
-        if isinstance(input.transform, InputTransform):
-            for state in input.transform._state.values():
-                data_pipeline_state.set_state(state)
+        for properties in [
+            self._train_input,
+            self._val_input,
+            self._test_input,
+            self._predict_input,
```
Comment (tchaton) [inline flash/core/data/new_data_module.py:304]: Let's also add the transforms there or at least ensure broadcasting for the input to its transform.

File: flash/core/integrations/transformers/states.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    from transformers import AutoTokenizer
+
+
+@dataclass(unsafe_hash=True, frozen=True)
+class TransformersBackboneState(ProcessState):
+    """The ``TransformersBackboneState`` records the ``backbone`` in use by tasks which rely on Hugging Face
+    transformers."""
+
+    backbone: str
+    tokenizer_kwargs: Optional[Dict[str, Any]] = field(default_factory=dict, hash=False)
```
Comment (tchaton) [inline flash/core/integrations/transformers/states.py:31]: Neat !

File: flash/core/data/new_data_module.py [hunk 3]
```diff
@@ -329,145 +327,3 @@ def on_after_batch_transfer(self, batch: Any, dataloader_idx: int) -> Any:
...
-
-        if enum not in cls.inputs_registry.available_keys():
-            available_constructors = [f"from_{key.name.lower()}" for key in cls.inputs_registry.available_keys()]
-            raise MisconfigurationException(
-                f"The ``AutoContainer`` ``inputs_registry`` doesn't contain the associated {enum} "
-                f"HINT: Here are the available constructors {available_constructors}"
-            )
-
-    @classmethod
-    def register_input(cls, enum: Union[str, LightningEnum], input_cls: Type[Input]) -> None:
```
Comment (tchaton) [inline flash/core/data/new_data_module.py:407]: So we remove entirely the input registry on the DataModule, I think it makes sense IMO.

File: flash/text/seq2seq/core/data.py [hunk 4]
```diff
@@ -62,290 +62,121 @@ def __setstate__(self, state):
 
 class Seq2SeqInput(Input):
     @requires("text")
-    def __init__(
+    def load_data(
```
Comment (tchaton) [inline flash/text/seq2seq/core/data.py:65]: I would name this one, `BaseSeq2SeqInput` as I don't expect users to provide their HF dataset, but I might be wrong.

File: flash/text/seq2seq/core/data.py [hunk 5]
```diff
@@ -355,63 +186,278 @@ def __init__(self):
...
-        state.pop("_tokenizer")
-        return state
-
-    def __setstate__(self, state):
-        self.__dict__.update(state)
-        self._backbone = None
-        _ = self.tokenizer
-
 
 class Seq2SeqData(DataModule):
```
Comment (tchaton) [inline flash/text/seq2seq/core/data.py:196]: Really neat !

**Review by tchaton — APPROVED** [review]
Looks great !

File: flash/text/seq2seq/core/data.py (hunk 4 above)
Comment (ethanwharris) [reply flash/text/seq2seq/core/data.py]: Done

File: flash/core/data/new_data_module.py (hunk 1 above)
Comment (ethanwharris) [reply flash/core/data/new_data_module.py:304]: Done :smiley:

**Pull request merged**